samples/star/src/main/kotlin/com/slack/circuit/star/repo/PetRepository.kt

```diff
@@ -2,49 +2,176 @@
+
+        // Re-populate the DB
+        for ((index, animal) in animals.withIndex()) {
+          starDb.starQueries.updateAnimal(
+            DbAnimal(
+              id = animal.id,
+              sort = index.toLong(),
+              // Names are sometimes all caps
+              name =
+                animal.name.lowercase().replaceFirstChar {
+                  if (it.isLowerCase()) it.titlecase(Locale.US) else it.toString()
+                },
+              url = animal.url,
+              photoUrls = animal.photos.map { it.full }.toImmutableList(),
+              primaryPhotoUrl = animal.photos.firstOrNull()?.medium,
+              tags =
+                listOfNotNull(
+                    animal.colors.primary,
+                    animal.colors.secondary,
+                    animal.breeds.primary,
+                    animal.breeds.secondary,
+                    animal.gender,
+                    animal.size,
+                    animal.status
+                  )
+                  .toImmutableList(),
+              description = animal.description,
+              primaryBreed = animal.breeds.primary,
+              gender = Gender.valueOf(animal.gender.uppercase(Locale.US)),
+              size = Size.valueOf(animal.size.uppercase(Locale.US)),
+              age = animal.age,
+            )
+          )
+        }
+
+        logUpdate("animals")
       }
+    } catch (e: HttpException) {
+      // Sometimes petfinder's API throws 429s for no reason.
+      // TODO retry?
+    }
   }
 
   override suspend fun getAnimalBio(id: Long): String? {
     val animal = getAnimal(id) ?: return null
-    return petFinderApi.animalBio(animal.url)
+    val opId = "bio:$id"
+    val isStale = isOperationStale(opId)
+    val dbBio by lazy(NONE) { starDb.starQueries.getAnimalBio(id).executeAsOneOrNull() }
+    return if (isStale || dbBio == null) {
+      petFinderApi.animalBio(animal.url).also { bio ->
+        // Single transaction to log the operation update with the put
+        starDb.transactionWithResult {
+          logUpdate(opId)
+          starDb.starQueries.putAnimalBio(AnimalBio(id, bio))
+        }
+      }
+    } else {
+      dbBio?.description
+    }
+  }
+
+  private fun logUpdate(operation: String) {
+    val timestamp = currentTimestamp()
+    starDb.starQueries.putUpdate(OpJournal(timestamp = timestamp, operation = operation))
+  }
+
+  /**
+   * Returns whether or not it's been more than one day since the last update to the given
+   * [operation].
+   */
+  private fun isOperationStale(operation: String): Boolean {
+    val lastUpdate =
+      starDb.starQueries.lastUpdate(operation).executeAsOneOrNull()?.timestamp ?: return true
+    val timestamp = currentTimestamp()
+    return Duration.between(Instant.ofEpochSecond(timestamp), Instant.ofEpochSecond(lastUpdate))
+      .toDays() > 1
+  }
+
+  private fun currentTimestamp(): Long {
+    return Instant.now().atZone(ZoneOffset.UTC).toEpochSecond()
```

Comment:
As I remembered, we can access `epochSecond` directly from Instant, wsithout coverting it to ZonedDateTime
```suggestion
    return Instant.now().epochSecond
```

Reply:
Nice, pushed in a separate commit as I thought it needed an import for some reason